### PR TITLE
Storage stage

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Validators need to check the work of replicators and also provide the interfaces that replicators will use to determine which key they should use to encrypt the ledger and which slice.

#### Summary of Changes

Add a storage stage which consumes entries and generates the encryption keys/results to validate replicator interactions.

Fixes #

### CODE REVIEW

File: src/storage_stage.rs [hunk 1]
```diff
@@ -0,0 +1,451 @@
+// A stage that handles generating the keys used to encrypt the ledger and sample it
+// for storage mining. Replicators submit storage proofs, validator then bundles them
+// to submit it's proof for mining to be rewarded.
```
Comment (garious) [inline src/storage_stage.rs]: its

File: src/storage_stage.rs [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+impl StorageStage {
+    pub fn new(
+        storage_state: &StorageState,
+        storage_entry_receiver: EntryReceiver,
+        ledger_path: Option<&str>,
+        keypair: Arc<Keypair>,
+        exit: Arc<AtomicBool>,
+        entry_height: u64,
+    ) -> Self {
+        let storage_keys_ = storage_state.storage_keys.clone();
```
Comment (garious) [inline src/storage_stage.rs]: Looks like these `clone()`s could be moved inside `spawn()` before `loop`.

File: src/storage_stage.rs [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+        let storage_results_ = storage_state.storage_results.clone();
+        let ledger_path = if let Some(ledger_str) = ledger_path {
```
Comment (garious) [inline src/storage_stage.rs]: I'm surprised clippy didn't suggest this, but how about?

```rust
let ledger_path = ledger_path.map(String::from);
```

File: src/storage_stage.rs [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+                _ledger_path,
+                slice as u64,
+                &mut storage_keys,
+                &samples,
+            ) {
+                Ok(hashes) => {
+                    debug!("Success! encrypted ledger slice: {}", slice);
+                    storage_results.copy_from_slice(&hashes);
+                }
+                Err(e) => info!("error encrypting file: {:?}", e),
```
Comment (garious) [inline src/storage_stage.rs]: Why not return this error?

File: src/storage_stage.rs [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+        ledger_path: &str,
+        poh_height: &mut u64,
+        entry_height: &mut u64,
+        current_key_idx: &mut usize,
+    ) -> Result<()> {
+        let timeout = Duration::new(1, 0);
+        let entries = entry_receiver.recv_timeout(timeout)?;
+
+        for entry in entries {
+            {
```
Comment (garious) [inline src/storage_stage.rs]: What do you get from this extra level of scope?

File: src/storage_stage.rs [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+                // Go through the transactions, find votes, and use them to update
+                // the storage_keys with their signatures.
+                for tx in entry.transactions {
```
Comment (garious) [inline src/storage_stage.rs]: Are you able to leverage `votes()` for this? https://github.com/solana-labs/solana/blob/master/src/ledger.rs#L498

File: src/storage_stage.rs [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+                }
+            }).unwrap();
+
+        StorageStage {
+            t_storage_mining_verifier,
+        }
+    }
+
+    pub fn process_entry_crossing(
+        _storage_results: &Arc<RwLock<StorageResults>>,
```
Comment (garious) [inline src/storage_stage.rs]: Why include the unused parameters?

Comment (sakridge) [reply src/storage_stage.rs]: They are used when `feature = 'cuda'` is set, but unused otherwise.

File: src/storage_stage.rs (hunk 5 above)
Comment (sakridge) [reply src/storage_stage.rs]: I was taking the lock there, but I moved it to inside the loop and forgot to remove it.

File: src/storage_stage.rs (hunk 6 above)
Comment (sakridge) [reply src/storage_stage.rs]: It doesn't give the signature, just (pubkey, vote, hash)

**Comment by pgarg66** [comment]
Seems, tip is broken..

```

--> src/storage_stage.rs:6:5
--
  | \|
  | 6 \| use chacha_cuda::chacha_cbc_encrypt_file_many_keys;
  | \|     ^^^^^^^^^^^ Maybe a missing `extern crate chacha_cuda;`?
  |  
  | error[E0277]: the size for values of type `[hash::Hash]` cannot be known at compilation time
  | --> src/storage_stage.rs:185:20
  | \|
  | 185 \|                 Ok(hashes) => {
  | \|                    ^^^^^^ doesn't have a size known at compile-time
  | \|
  | = help: the trait `std::marker::Sized` is not implemented for `[hash::Hash]`
  | = note: to learn more, visit <https://doc.rust-lang.org/book/second-edition/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
  | = note: all local variables must have a statically known size
  | = help: unsized locals are gated as an unstable feature
  |  
  | error[E0277]: the size for values of type `[hash::Hash]` cannot be known at compilation time
  | --> src/storage_stage.rs:185:17
  | \|
  | 185 \|                 Ok(hashes) => {
  | \|                 ^^^^^^^^^^ doesn't have a size known at compile-time
  | \|
  | = help: the trait `std::marker::Sized` is not implemented for `[hash::Hash]`
  | = note: to learn more, visit <https://doc.rust-lang.org/book/second-edition/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
  | = note: required by `serde::export::Ok`
  |  
  | error: aborting due to 3 previous errors
  |  

```

**Pull request merged**